core/services/mongodb-remote/src/test/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizerUnitTests.kt

```diff
@@ -1,190 +1,882 @@
+        // and sync'd locally
+        assertEquals(
+            docAfterUpdate,
+            ctx.findTestDocumentFromLocalCollection())
+
+        // 3: Update -> Conflict -> Exception -> Freeze
+        // reset (delete, insert, sync)
+        ctx = harness.freshTestContext()
+        docAfterUpdate = BsonDocument("count", BsonInt32(2)).append("_id", ctx.testDocumentId)
+        expectedLocalEvent = ChangeEvent.changeEventForLocalUpdate(
+            ctx.namespace,
+            ctx.testDocumentId,
+            ctx.updateDocument,
+            docAfterUpdate,
+            true)
+        expectedRemoteEvent = ChangeEvent.changeEventForLocalDelete(ctx.namespace, ctx.testDocumentId, false)
+
+        ctx.mockUpdateResult(RemoteUpdateResult(0, 0, null))
+
+        ctx.insertTestDocument()
+        ctx.waitForEvent()
+        ctx.doSyncPass()
+        ctx.waitForEvent()
+        ctx.verifyChangeEventListenerCalledForActiveDoc(
+            1, ChangeEvent.changeEventForLocalInsert(ctx.namespace, ctx.testDocument, false))
+        ctx.doSyncPass()
+
+        // update the reset doc
+        ctx.updateTestDocument()
+        ctx.waitForEvent()
+
+        // prepare an exceptionToThrow to be thrown, and sync
+        ctx.exceptionToThrowDuringConflict = Exception("bad")
+        ctx.doSyncPass()
+        ctx.waitForError()
+
+        // verify that, though the conflict handler was called, the exceptionToThrow was emitted
+        // by the dataSynchronizer
+        ctx.verifyChangeEventListenerCalledForActiveDoc(times = 0)
+        ctx.verifyConflictHandlerCalledForActiveDoc(1, expectedLocalEvent, expectedRemoteEvent)
+        ctx.verifyErrorListenerCalledForActiveDoc(1, ctx.exceptionToThrowDuringConflict)
+
+        // assert that this document is still the locally updated doc. this is frozen now
+        assertEquals(docAfterUpdate, ctx.findTestDocumentFromLocalCollection())
+
+        // clear issues. open a path for a delete.
+        // do another sync pass. the doc should remain the same as it is frozen
+        ctx.exceptionToThrowDuringConflict = null
+        ctx.shouldConflictBeResolvedByRemote = false
+
+        ctx.doSyncPass()
+        assertEquals(docAfterUpdate, ctx.findTestDocumentFromLocalCollection())
+
+        // update the doc locally, unfreezing it, and syncing it
+        ctx.mockUpdateResult(RemoteUpdateResult(1, 1, null))
+        assertEquals(1L, ctx.updateTestDocument().matchedCount)
+        ctx.doSyncPass()
+
+        // 4: Unknown -> Freeze
+        ctx = harness.freshTestContext()
+        ctx.insertTestDocument()
+        ctx.queueConsumableRemoteUnknownEvent()
+        ctx.doSyncPass()
+        assertEquals(ctx.testDocument, ctx.findTestDocumentFromLocalCollection())
+
+        // should be frozen since the operation type was unknown
+        ctx.queueConsumableRemoteUpdateEvent()
+        ctx.doSyncPass()
+        assertEquals(ctx.testDocument, ctx.findTestDocumentFromLocalCollection())
```

Comment:
[nit q] same q as before, how do we know this is frozen? don't we need to try deleting first?

Reply:
We don't need to try deleting– making sure the document doesn't update works as well.